File: typescript/base/src/main/vsdk/toolkit/environment/scene/SimpleScene.ts
// Java basic classes
import { ArrayList } from "../../../../java/util/ArrayList.js";

// VSDK Classes
import { Entity } from "../../common/Entity.js";
import type { Background } from "../background/Background.js";
import type { Camera } from "../camera/Camera.js";
import type { CameraSnapshot } from "../camera/CameraSnapshot.js";
import type { Light } from "../light/Light.js";
import type { SimpleBody } from "./SimpleBody.js";
import { SimpleSceneSnapshot } from "./SimpleSceneSnapshot.js";

export class SimpleScene extends Entity {
    private simpleBodies: ArrayList<SimpleBody>;
    private lights: ArrayList<Light>;
    private backgrounds: ArrayList<Background>;
    private cameras: ArrayList<Camera>;
    private activeCameraIndex = 0;
    private activeBackgroundIndex = 0;

    public constructor() {
        super();
        this.simpleBodies = new ArrayList<SimpleBody>();
        this.lights = new ArrayList<Light>();
        this.backgrounds = new ArrayList<Background>();
        this.cameras = new ArrayList<Camera>();
    }

    public getActiveCameraIndex(): number {
        return this.activeCameraIndex;
    }

    public getActiveBackgroundIndex(): number {
        return this.activeBackgroundIndex;
    }

    public setActiveCameraIndex(i: number): void {
        this.activeCameraIndex = i;
    }

    public setActiveBackgroundIndex(i: number): void {
        this.activeBackgroundIndex = i;
    }

    public addBody(b: SimpleBody): void {
        this.simpleBodies.add(b);
    }

    public addCamera(c: Camera): void {
        this.cameras.add(c);
    }

    public addBackground(b: Background): void {
        this.backgrounds.add(b);
    }

    public addLight(l: Light): void {
        l.setId(this.lights.size());
        this.lights.add(l);
    }

    public getSimpleBodies(): ArrayList<SimpleBody> {
        return this.simpleBodies;
    }

    public getLights(): ArrayList<Light> {
        return this.lights;
    }

    public getBackgrounds(): ArrayList<Background> {
        return this.backgrounds;
    }

    public getCameras(): ArrayList<Camera> {
        return this.cameras;
    }

    public setSimpleBodies(simpleBodies: ArrayList<SimpleBody>): void {
        this.simpleBodies = simpleBodies;
    }

    public setLights(lights: ArrayList<Light>): void {
        this.lights = lights;
        for (let i = 0; i < this.lights.size(); i++) {
            this.lights.get(i).setId(i);
        }
    }

    public setBackgrounds(backgrounds: ArrayList<Background>): void {
        this.backgrounds = backgrounds;
    }

    public getActiveBackground(): Background {
        return this.backgrounds.get(this.activeBackgroundIndex);
    }

    public getActiveCamera(): Camera {
        return this.cameras.get(this.activeCameraIndex);
    }

    public setCameras(cameras: ArrayList<Camera>): void {
        this.cameras = cameras;
    }

    public exportToSimpleSceneSnapshot(): SimpleSceneSnapshot;
    public exportToSimpleSceneSnapshot(viewportXSize: number, viewportYSize: number): SimpleSceneSnapshot;
    public exportToSimpleSceneSnapshot(cameraSnapshot: CameraSnapshot, background: Background): SimpleSceneSnapshot;
    public exportToSimpleSceneSnapshot(a?: number | CameraSnapshot, b?: number | Background): SimpleSceneSnapshot {
        if (a === undefined) {
            return this.exportToSimpleSceneSnapshot(
                this.getActiveCamera().exportToCameraSnapshot(),
                this.getActiveBackground(),
            );
        }
        if (typeof a === "number") {
            return this.exportToSimpleSceneSnapshot(
                this.getActiveCamera().exportToCameraSnapshot(a, b as number),
                this.getActiveBackground(),
            );
        }
        return new SimpleSceneSnapshot(this.simpleBodies, this.lights, b as Background, a);
    }
}

File: java/base/src/main/vsdk/toolkit/environment/scene/SimpleScene.java
package vsdk.toolkit.environment.scene;
import java.io.Serial;

// Java basic classes
import java.util.ArrayList;

// VSDK Classes
import vsdk.toolkit.common.Entity;
import vsdk.toolkit.environment.background.Background;
import vsdk.toolkit.environment.camera.Camera;
import vsdk.toolkit.environment.camera.CameraSnapshot;
import vsdk.toolkit.environment.light.Light;

public class SimpleScene extends Entity
{
    @Serial private static final long serialVersionUID = 20100901L;

    private ArrayList<SimpleBody> simpleBodies;
    private ArrayList<Light> lights;
    private ArrayList<Background> backgrounds;
    private ArrayList<Camera> cameras;
    private int activeCameraIndex;
    private int activeBackgroundIndex;

    public SimpleScene()
    {
        simpleBodies = new ArrayList<SimpleBody>();
        lights = new ArrayList<Light>();
        backgrounds = new ArrayList<Background>();
        cameras = new ArrayList<Camera>();
    }

    public int getActiveCameraIndex()
    {
        return activeCameraIndex;
    }

    public int getActiveBackgroundIndex()
    {
        return activeBackgroundIndex;
    }

    public void setActiveCameraIndex(int i)
    {
        activeCameraIndex = i;
    }

    public void setActiveBackgroundIndex(int i)
    {
        activeBackgroundIndex = i;
    }

    public void addBody(SimpleBody b)
    {
        simpleBodies.add(b);
    }

    public void addCamera(Camera c)
    {
        cameras.add(c);
    }

    public void addBackground(Background b)
    {
        backgrounds.add(b);
    }

    public void addLight(Light l)
    {
        l.setId(lights.size());
        lights.add(l);
    }

    public ArrayList<SimpleBody> getSimpleBodies()
    {
        return simpleBodies;
    }

    public ArrayList<Light> getLights()
    {
        return lights;
    }

    public ArrayList<Background> getBackgrounds()
    {
        return backgrounds;
    }

    public ArrayList<Camera> getCameras()
    {
        return cameras;
    }

    public void setSimpleBodies(ArrayList<SimpleBody> simpleBodies)
    {
        this.simpleBodies = simpleBodies;
    }

    public void setLights(ArrayList<Light> lights)
    {
        this.lights = lights;
        for ( int i = 0; i < this.lights.size(); i++ ) {
            this.lights.get(i).setId(i);
        }
    }

    public void setBackgrounds(ArrayList<Background> backgrounds)
    {
        this.backgrounds = backgrounds;
    }

    public Background getActiveBackground()
    {
        return backgrounds.get(activeBackgroundIndex);
    }

    public Camera getActiveCamera()
    {
        return cameras.get(activeCameraIndex);
    }

    public void setCameras(ArrayList<Camera> cameras)
    {
        this.cameras = cameras;
    }

    public SimpleSceneSnapshot exportToSimpleSceneSnapshot()
    {
        return exportToSimpleSceneSnapshot(
            getActiveCamera().exportToCameraSnapshot(),
            getActiveBackground());
    }

    public SimpleSceneSnapshot exportToSimpleSceneSnapshot(
        int viewportXSize,
        int viewportYSize)
    {
        return exportToSimpleSceneSnapshot(
            getActiveCamera().exportToCameraSnapshot(viewportXSize, viewportYSize),
            getActiveBackground());
    }

    public SimpleSceneSnapshot exportToSimpleSceneSnapshot(
        CameraSnapshot cameraSnapshot,
        Background background)
    {
        return new SimpleSceneSnapshot(
            simpleBodies,
            lights,
            background,
            cameraSnapshot);
    }
}
